PluginManager: Dont leave leftover files on unsuccessful installs

spinscale opened the pull request:
If the plugin manager cannot successfully install a plugin, ensure
that every directory is cleaned up again. This includes

plugins/foo
config/foo
bin/foo

Closes #12749

dadoonet commented on core/src/main/java/org/elasticsearch/plugins/PluginManager.java:
```diff
@@ -208,17 +218,53 @@ private void extract(PluginHandle pluginHandle, Terminal terminal, Path pluginFi
         IOUtils.rm(tmp, pluginFile);
```
May be here use `tryToDeletePath()` instead?

spinscale replied:
yeah, did that

dadoonet commented on core/src/test/java/org/elasticsearch/plugins/PluginManagerPermissionTests.java:
```diff
@@ -0,0 +1,295 @@
+
+    // TODO ONLY RUN IF FILESYSTEM PERMISSIONS WORK
+    public void testThatUnaccessibleBinDirectoryAbortsPluginInstallation() throws Exception {
+        assumeTrue("File system does not support permissions, skipping", supportsPermissions);
```
I'm unsure if it works but as all your tests require FS to support permissions, may be you can move this `assumeTrue` in `setup()`?

spinscale replied:
note that not all tests require this

dadoonet commented:
Left 2 small comments. It looks good to me.

clintongormley commented on core/src/main/java/org/elasticsearch/plugins/PluginManager.java:
```diff
@@ -183,6 +183,16 @@ private void extract(PluginHandle pluginHandle, Terminal terminal, Path pluginFi
         }
 
+        Path pluginConfigDirectory = pluginHandle.configDir(environment);
+        if (Files.exists(pluginConfigDirectory) && !Files.isDirectory(pluginConfigDirectory)) {
```
Should we be checking if the directory is writable here?

spinscale replied:
not needed, one test is covering this already and returning an appropriate exception

spinscale commented:
fixed the review comments, rebased against master with a fair share of changes and only run the logic to copy `bin` and `config` dirs if actually needed, otherwise exit early

please give it another look

dadoonet commented on core/src/main/java/org/elasticsearch/plugins/PluginManager.java:
```diff
@@ -241,22 +240,75 @@ private void extract(PluginHandle pluginHandle, Terminal terminal, Path pluginFi
+
+        Path sourceConfigDirectory = extractLocation.resolve("config");
+        Path destConfigDirectory = pluginHandle.configDir(environment);
+        boolean needToCopyConfigDirectory = Files.exists(sourceConfigDirectory);
```
I guess you create this `needToCopyConfigDirectory` instead of adding documentation, right?
If not, may be:

```
if (Files.exists(sourceConfigDirectory)) {
```

dadoonet replied:
Also, we were testing in the past that `sourceConfigDirectory` is actually a dir and not only an existing file.
Did you change that on purpose? Or should it be `if (Files.isDirectory(sourceConfigDirectory)) {`

spinscale replied:
yes, I prefer well named variables instead of comments. In this example though it is not too useful. There used to be two conditions, but now that it is only one, it's not that helpful anymore

dadoonet commented on core/src/main/java/org/elasticsearch/plugins/PluginManager.java:
```diff
@@ -241,22 +240,75 @@ private void extract(PluginHandle pluginHandle, Terminal terminal, Path pluginFi
+        Path destConfigDirectory = pluginHandle.configDir(environment);
+        boolean needToCopyConfigDirectory = Files.exists(sourceConfigDirectory);
+        if (needToCopyConfigDirectory) {
+            if (Files.exists(destConfigDirectory) && !Files.isDirectory(destConfigDirectory)) {
```
Or may be you the test is wrong here and you need to test the `sourceConfigDirectory` instead?

dadoonet replied:
May be add a test if sourceConfigDir is actually a file and not a dir?
Same for bin?

I know just corner cases... :)

nik9000 replied:
That's certainly a funky corner case, right? Its a broken plugin that does that.

dadoonet commented on core/src/main/java/org/elasticsearch/plugins/PluginManager.java:
```diff
@@ -241,22 +240,75 @@ private void extract(PluginHandle pluginHandle, Terminal terminal, Path pluginFi
+            }
+
+            try {
+                copyBinDirectory(sourcePluginBinDirectory, pluginHandle, terminal);
```
Why not `copyBinDirectory(sourcePluginBinDirectory, destPluginBinDirectory, terminal);` instead?

As far as I can see you pass `pluginHandle` just for logging reason here. You could pass the plugin name instead:

```java
copyBinDirectory(sourcePluginBinDirectory, destPluginBinDirectory, pluginHandle.name, terminal);
```

dadoonet commented:
@spinscale I left some comments.

nik9000 commented on core/src/main/java/org/elasticsearch/plugins/PluginManager.java:
```diff
@@ -241,22 +240,75 @@ private void extract(PluginHandle pluginHandle, Terminal terminal, Path pluginFi
+        if (needToCopyBinDirectory) {
+            if (Files.exists(destPluginBinDirectory) && !Files.isDirectory(destPluginBinDirectory)) {
+                tryToDeletePath(terminal, extractLocation);
+                throw new IOException("plugin bin directory " + destPluginBinDirectory + " is not a directory");
```
Do you mean to fail here after deleting the path? Maybe WARN and then delete the path and fail if we can't delete it?

nik9000 replied:
Ah! I see now. You don't do mean to fail. You aren't deleting the destPluginBinDirectory, you are deleting the extractLocation to clean up. Got it! +1

nik9000 commented:
@spinscale, there are a few comments around more you could do here but I think its an improvement as is. Its worth merging, I think.

I love the tests. Very descriptive method names help a ton.

If you want to add the tests for sourceConfigPath being a file instead of a dir I'd cause that to be an "abort fast" thing. Its a pretty broken plugin that does it.

spinscale commented:
@dadoonet updated your last comment and rebased against master, mind to take another view?

nik9000 commented:
LGTM
